src/Compiler/Checking/CheckExpressions.fs

```diff
@@ -5692,6 +5692,19 @@ and TcExprThen cenv (overallTy: OverallTy) env tpenv isArg synExpr delayed =
             TcExprThen cenv overallTy env tpenv isArg (SynExpr.LongIdent (isOpt, SynLongIdent([altId], [], [None]), None, mLongId)) delayed
         | _ -> TcLongIdentThen cenv overallTy env tpenv longId delayed
 
+    // x?k<-42
+    | SynExpr.Set(SynExpr.Dynamic(e1, _, e2, _) , rhsExpr, m) ->
+        let e2 =
```

Comment:
We should factor all the branches of this function into separate functions, like I've been doing for TcExpr and TcPat.  I guess you may as well start with this one